# SpotiScry
### Date: January 2021
### Project Goal:
### Creators:

## Table of Contents
- [Acquire](#acquire)
- [Prepare](#prepare)
- [Explore](#explore)
- [Model](#model)
- [Conclusions](#conclusions)

# Acquire <a class="anchor" id="acquire"></a>

## Takeaways

# Prepare <a class="anchor" id="prepare"></a>

## Takeaways

# Explore <a class="anchor" id="explore"></a>

## Takeaways

# Model <a class="anchor" id="model"></a>

### Prepare data for modeling

In [1]:
from preprocessing import spotify_split, scale_data, modeling_prep
from model import get_baseline_metrics, linear_regression_model, lasso_lars
from model import polynomial_regression, evaluate_df

In [2]:
df = modeling_prep()

In [3]:
# split the data for modeling
X_train, y_train, X_validate, y_validate, X_test, y_test, train, validate, test = spotify_split(df, 'popularity')
X_train.head(3)

Shape of train: (4012, 24) | Shape of validate: (861, 24) | Shape of test: (860, 24)
Percent train: 70.0        | Percent validate: 15.0       | Percent test: 15.0


danceability  energy  key  loudness  mode  \
track_id                                                            
30bqVoKjX479ab90a8Pafp         0.585   0.471    4    -9.934     0   
0HO8pCseEpgozNi3z0R4bc         0.833   0.518   10   -10.126     0   
643K3eEgRvdJiXjSzlz7dg         0.471   0.671    1    -6.050     1   

                        speechiness  instrumentalness  liveness  valence  \
track_id                                                                   
30bqVoKjX479ab90a8Pafp       0.0616            0.0184     0.115    0.323   
0HO8pCseEpgozNi3z0R4bc       0.3490            0.0000     0.635    0.773   
643K3eEgRvdJiXjSzlz7dg       0.3410            0.0000     0.308    0.850   

                          tempo  ...  duration_seconds  duration_minutes  \
track_id                         ...                                       
30bqVoKjX479ab90a8Pafp   93.099  ...               142                 2   
0HO8pCseEpgozNi3z0R4bc  180.008  ...               120                 2   
643K3eEgRvdJiXjSzlz7dg  176.863  ...               252                 4   

                        is_featured_artist  release_year  release_month  \
track_id                                                                  
30bqVoKjX479ab90a8Pafp                   0          2019              4   
0HO8pCseEpgozNi3z0R4bc                   0          2015              3   
643K3eEgRvdJiXjSzlz7dg                   0          1998              8   

                        release_day  is_explicit  top_ten_label  compilation  \
track_id                                                                       
30bqVoKjX479ab90a8Pafp           19            1              0            0   
0HO8pCseEpgozNi3z0R4bc           17            1              0            0   
643K3eEgRvdJiXjSzlz7dg           11            1              0            0   

                        single  
track_id                        
30bqVoKjX479ab90a8Pafp       1  
0HO8pCseEpgozNi3z0R4bc       0  
643K3eEgRvdJiXjSzlz7dg       0  

[3 rows x 24 columns]

In [4]:
# scale the data: min-max scaler
X_train_scaled, X_validate_scaled, X_test_scaled = scale_data(train, validate, test, 'popularity', 'MinMax')
X_train_scaled.head(3)

danceability    energy       key  loudness  mode  \
track_id                                                                   
30bqVoKjX479ab90a8Pafp      0.593306  0.439493  0.363636  0.580823   0.0   
0HO8pCseEpgozNi3z0R4bc      0.844828  0.489960  0.909091  0.572667   0.0   
643K3eEgRvdJiXjSzlz7dg      0.477688  0.654247  0.090909  0.745826   1.0   

                        speechiness  instrumentalness  liveness   valence  \
track_id                                                                    
30bqVoKjX479ab90a8Pafp     0.064033          0.019127  0.101302  0.328921   
0HO8pCseEpgozNi3z0R4bc     0.362786          0.000000  0.651741  0.787169   
643K3eEgRvdJiXjSzlz7dg     0.354470          0.000000  0.305600  0.865580   

                           tempo  ...  duration_seconds  duration_minutes  \
track_id                          ...                                       
30bqVoKjX479ab90a8Pafp  0.422695  ...          0.235832          0.222222   
0HO8pCseEpgozNi3z0R4bc  0.817286  ...          0.195612          0.222222   
643K3eEgRvdJiXjSzlz7dg  0.803007  ...          0.436929          0.444444   

                        is_featured_artist  release_year  release_month  \
track_id                                                                  
30bqVoKjX479ab90a8Pafp                 0.0      0.972973       0.272727   
0HO8pCseEpgozNi3z0R4bc                 0.0      0.864865       0.181818   
643K3eEgRvdJiXjSzlz7dg                 0.0      0.405405       0.636364   

                        release_day  is_explicit  top_ten_label  compilation  \
track_id                                                                       
30bqVoKjX479ab90a8Pafp     0.600000          1.0            0.0          0.0   
0HO8pCseEpgozNi3z0R4bc     0.533333          1.0            0.0          0.0   
643K3eEgRvdJiXjSzlz7dg     0.333333          1.0            0.0          0.0   

                        single  
track_id                        
30bqVoKjX479ab90a8Pafp     1.0  
0HO8pCseEpgozNi3z0R4bc     0.0  
643K3eEgRvdJiXjSzlz7dg     0.0  

[3 rows x 24 columns]

### Baseline Model

In [5]:
#baseline is the mean popularity of the training dataset
bl, bl_train_rmse = get_baseline_metrics(y_train)

RMSE (Root Mean Square Error) of Baseline on train data:
 22.897138


### Train Models

In [13]:
# OLS Model(defaults)
lm_rmse, lm_rmse_v = linear_regression_model(X_train, y_train, X_validate, y_validate)
# LASSO + LARS Model(alpha=.00001)
lars_rmse, lars_rmse_v, lars_rmse_t = lasso_lars(X_train, y_train, X_validate, y_validate, X_test, y_test, alpha=.00001, fit_intercept=True)
# Polynomial Features (squared, deg=2) with Linear Regression
lm_sq_rmse, lm_sq_rmse_v = polynomial_regression(X_train, y_train, X_validate, y_validate, 'Squared', degree=2) 
# Polynomial Features (cubed, deg=3) with Linear Regression
lm_cb_rmse, lm_cb_rmse_v =  polynomial_regression(X_train, y_train, X_validate, y_validate, 'Cubed', degree=3) 

RMSE for OLS using Linear Regression 

On train data:
 20.910148 

RMSE for LASSO + LARS 

On train data:
 20.911551 

RMSE for Polynomial Squared + Linear Regression 

On train data:
 19.015856 

RMSE for Polynomial Cubed + Linear Regression 

On train data:
 17.552747 



### Evaluate Models

In [14]:
evaluate_df(bl_train_rmse, lm_rmse, lars_rmse, lars_rmse_v, 
            lars_rmse_t, lm_sq_rmse, lm_sq_rmse_v, 
            lm_cb_rmse, lm_cb_rmse_v)

Model beat baseline by 9.22%


train_rmse validate_rmse test_rmse
pf3_lr      17.552747       45.4758         -
pf2_lr      19.015856       21.2229         -
ols         20.910148             -         -
lassolars   20.911551        20.891   20.7853
baseline    22.897138             -         -

## Takeaways

<div class="alert alert-block alert-info">
    <li>Polynomial features cubed is overfit</li>
    <li>Polynomial features squared and Lasso lars perform best on train and validate</li>
    <li><b>Best Model</b> is Lasso Lars since it performed better on unseen data</li>
    <li>Beat the baseline by 9.22%.</li>
    </div>

# Conclusions <a class="anchor" id="conclusions"></a>